DaisyUI v5 + Tailwind v4 + RJSF v6

inchoate opened the pull request:
### Reasons for making this change

[Please describe them here]

If this is related to existing tickets, include links to them as well. Use the syntax `fixes #[issue number]` (ex: `fixes #123`).

If your PR is non-trivial and you'd like to schedule a synchronous review, please add it to the weekly meeting agenda: https://docs.google.com/document/d/12PjTvv21k6LIky6bNQVnsplMLLnmEuypTLQF8a-8Wss/edit

### Checklist

- [ ] **I'm updating documentation**
  - [ ] I've [checked the rendering](https://rjsf-team.github.io/react-jsonschema-form/docs/contributing) of the Markdown text I've added
- [ ] **I'm adding or updating code**
  - [ ] I've added and/or updated tests. I've run `npx nx run-many --target=build --exclude=@rjsf/docs && npm run test:update` to update snapshots, if needed.
  - [ ] I've updated [docs](https://rjsf-team.github.io/react-jsonschema-form/docs) if needed
  - [ ] I've updated the [changelog](https://github.com/rjsf-team/react-jsonschema-form/blob/main/CHANGELOG.md) with a description of the PR
- [ ] **I'm adding a new feature**
  - [ ] I've updated the playground with an example use of the feature

inchoate commented:
Copy of https://github.com/rjsf-team/react-jsonschema-form/pull/4495 for RJSF v6. https://github.com/rjsf-team/react-jsonschema-form/pull/4495

nickgros commented:
@heath-freenome Can you provide guidance to @inchoate about your new grid templating system in v6 and what other major changes (if any) he might expect to impact this PR? Would it make sense for us to look at prioritizing adding this and #4497 to v6 before additional templating changes?

inchoate commented:
Forgot the v6 screenshots:

**Themes and Subthemes**
<img width="1031" alt="Screenshot 2025-02-14 at 4 11 11 PM" src="https://github.com/user-attachments/assets/0279c663-2124-4ba1-a3ed-36f369cf2750" />
<img width="619" alt="Screenshot 2025-02-14 at 4 11 28 PM" src="https://github.com/user-attachments/assets/e5d89d7f-8b66-4922-84bc-4d626669a620" />

**Simple Form in the Retro and Dark Themes**
<img width="502" alt="Screenshot 2025-02-14 at 4 11 54 PM" src="https://github.com/user-attachments/assets/884842d4-cc8b-41bf-9b3d-fe1bb264c2b7" />
<img width="510" alt="Screenshot 2025-02-14 at 4 12 20 PM" src="https://github.com/user-attachments/assets/0ba68263-3587-4d34-9ce3-6db81ee00fea" />

**Numbers in the Pastel Theme**
<img width="522" alt="Screenshot 2025-02-14 at 4 13 00 PM" src="https://github.com/user-attachments/assets/96e6fdf3-2915-4368-9538-898f55d57eea" />

**File Uploaders in the caramellatte Theme**
- supports single, multiple, and type filters
<img width="523" alt="Screenshot 2025-02-14 at 4 13 45 PM" src="https://github.com/user-attachments/assets/e53f7b35-1eef-4eec-bfcf-66178b0f9dac" />

**Datetime picker in the Cupcake, Dracula, and  CMYK Themes**
- integrates w/react-date-picker
<img width="506" alt="Screenshot 2025-02-14 at 4 14 59 PM" src="https://github.com/user-attachments/assets/eba372fe-732c-4695-a477-7cf3f0e33735" />
<img width="494" alt="Screenshot 2025-02-14 at 4 17 28 PM" src="https://github.com/user-attachments/assets/9a6c5016-54fb-4b81-bfdd-9b1b4b50415c" />
<img width="519" alt="Screenshot 2025-02-14 at 4 18 03 PM" src="https://github.com/user-attachments/assets/9efdb4ed-955f-473d-af3b-a3af73c84ede" />

**Enumerated Objects in the Light Theme**
<img width="526" alt="Screenshot 2025-02-14 at 4 16 07 PM" src="https://github.com/user-attachments/assets/5e06b595-0e99-41c0-932a-2c7c68c3f912" />

Hope this helps.

heath-freenome commented:
> @heath-freenome Can you provide guidance to @inchoate about your new grid templating system in v6 and what other major changes (if any) he might expect to impact this PR? Would it make sense for us to prioritize merging this and #4497 to v6 before additional templating changes?

@nickgros @inchoate I've completed all of the template changes in V6 already. It seems like there needs to be an implementation of `GridTemplate` for this theme implementation.

heath-freenome commented on packages/daisyui/src/templates/ArrayFieldItemTemplate/ArrayFieldItemTemplate.tsx:
```diff
@@ -0,0 +1,77 @@
+      <div className='card-body p-4'>
+        <div className='array-field-item-content mb-2'>{children}</div>
+        <div className='card-actions justify-end'>
+          {hasMoveUp && (
```
You will want to use the `ArrayFieldItemButtonsTemplate` in place of the buttons here. Although your implementation has additional behavior to the default implementation so maybe it is not necessary.

heath-freenome commented on packages/daisyui/src/daisyForm/DaisyForm.tsx:
```diff
@@ -0,0 +1,20 @@
+  F extends FormContextType = any
+>(): ComponentType<FormProps<T, S, F>> {
+  const theme = generateTheme<T, S, F>();
+  console.log('Generated theme:', theme); // Debug what templates are available
```
Please remove all of the `console.log()` statements before committing

heath-freenome commented on packages/daisyui/src/templates/ArrayFieldTitleTemplate/ArrayFieldTitleTemplate.tsx:
```diff
@@ -0,0 +1,12 @@
+  console.log('DaisyUI ArrayFieldTitleTemplate');
+
+  props;
+  return <div>ArrayFieldTitleTemplate</div>;
```
I'm curious why you implemented this since it does nothing to display the `TitleFieldTemplate`

heath-freenome commented on packages/daisyui/src/templates/BaseInputTemplate/index.ts:
```diff
@@ -0,0 +1,2 @@
+export { default } from './BaseInputTemplate';
+export * from './BaseInputTemplate';
```
```suggestion
export * from './BaseInputTemplate';

```

heath-freenome commented on packages/daisyui/src/templates/DescriptionField/index.ts:
```diff
@@ -0,0 +1,2 @@
+export { default } from './DescriptionField';
+export * from './DescriptionField';
```
```suggestion
export * from './DescriptionField';

```

heath-freenome commented on packages/daisyui/src/templates/ErrorList/index.ts:
```diff
@@ -0,0 +1,2 @@
+export { default } from './ErrorList';
+export * from './ErrorList';
```
```suggestion
export * from './ErrorList';

```

heath-freenome commented on packages/daisyui/src/templates/FieldTemplate/index.ts:
```diff
@@ -0,0 +1,2 @@
+export { default } from './FieldTemplate';
+export * from './FieldTemplate';
```
```suggestion
export * from './FieldTemplate';

```

heath-freenome commented on packages/daisyui/src/templates/ObjectFieldTemplate/index.ts:
```diff
@@ -0,0 +1,2 @@
+export { default } from './ObjectFieldTemplate';
+export * from './ObjectFieldTemplate';
```
```suggestion
export * from './ObjectFieldTemplate';

```

heath-freenome commented on packages/daisyui/src/templates/WrapIfAdditionalTemplate/index.ts:
```diff
@@ -0,0 +1,2 @@
+export { default } from './WrapIfAdditionalTemplate';
+export * from './WrapIfAdditionalTemplate';
```
```suggestion
export * from './WrapIfAdditionalTemplate';

```

heath-freenome commented on packages/daisyui/src/theme/index.ts:
```diff
@@ -0,0 +1,3 @@
+export { default } from './Theme';
+export * from './Theme';
+
```
```suggestion
```

heath-freenome commented on packages/daisyui/src/types.d.ts:
```diff
@@ -0,0 +1,4 @@
+declare module '*.css?inline' {
+    const content: string;
+    export default content;
+} 
```
```suggestion
} 

```

inchoate commented on packages/daisyui/src/utils.ts:
```diff
@@ -0,0 +1,31 @@
+export function getDaisy({ uiSchema = {} }: GetDaisyProps): DaisyProps {
+    const daisyProps = (uiSchema['ui:options'] && uiSchema['ui:options'].daisy) || {};
+
+    // Object.keys(daisyProps).forEach((key) => {
+    //     /**
+    //      * Leveraging `shouldForwardProp` to remove props
+    //      *
+    //      * This is a utility function that's used in `@daisy-ui/react`'s factory function.
+    //      * Normally, it prevents DaisyProps from being passed to the DOM.
+    //      * In this case we just want to delete the unknown props. So we flip the boolean.
+    //      */
+    //     if (shouldForwardProp(key)) {
+    //         delete (daisyProps as any)[key];
+    //     }
+    // });
```
Removed.

inchoate commented:
@heath-freenome thanks for your comments. I believe I addressed everything you brought up. Clearly my commit hooks were not set up correctly and, yet they still refuse to run.

I also added:

**Toggle**
<img width="1175" alt="Screenshot 2025-02-18 at 4 01 26 PM" src="https://github.com/user-attachments/assets/39436929-fd1a-42c0-b312-2fe1e9da1e00" />

**Ratings**
<img width="1230" alt="Screenshot 2025-02-18 at 3 59 20 PM" src="https://github.com/user-attachments/assets/d54ca373-94cd-48d9-aa08-07684c6baf8a" />

@nickgros

heath-freenome commented:
@inchoate I don't see an implementation of the `GridTemplate`.

heath-freenome commented on packages/daisyui/src/templates/Templates.tsx:
```diff
@@ -0,0 +1,49 @@
+    FieldErrorTemplate,
+    FieldHelpTemplate,
+    FieldTemplate,
+    ObjectFieldTemplate,
```
You are missing a `GridTemplate` implementation.

inchoate commented:
> @inchoate I don't see an implementation of the `GridTemplate`.

Sorry. Didn't see you make a comment about that. I'll just read the core implementation and go from there unless I hear otherwise.

inchoate commented:
@nickgros / @heath-freenome I'm kinda' flailing on the GridTemplate right now. How can I accurately test a different theme's implementation on the v6 branch?

heath-freenome commented:
I'm working on a new Field that will use it. Give me a few days to get it committed

inchoate commented:
Awesome. Thanks.

On Wed, Feb 19, 2025, 12:16 PM Heath C ***@***.***> wrote:

> I'm working on a new Field that will use it. Give me a few days to get it
> committed
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/rjsf-team/react-jsonschema-form/pull/4498#issuecomment-2669419503>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAH34OUF3YWNPHYPFMIDJF32QTC6JAVCNFSM6AAAAABXE7EPRCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMNRZGQYTSNJQGM>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
> [image: heath-freenome]*heath-freenome* left a comment
> (rjsf-team/react-jsonschema-form#4498)
> <https://github.com/rjsf-team/react-jsonschema-form/pull/4498#issuecomment-2669419503>
>
> I'm working on a new Field that will use it. Give me a few days to get it
> committed
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/rjsf-team/react-jsonschema-form/pull/4498#issuecomment-2669419503>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAH34OUF3YWNPHYPFMIDJF32QTC6JAVCNFSM6AAAAABXE7EPRCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMNRZGQYTSNJQGM>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

heath-freenome commented:
I've merged a few PRs with updates and fixes from `main`. You'll have to rebase and update your PR. Also, checkout the new `Layout Grid` in the playground and update the `layoutGrid.tsx` in the `samples` directory to add your custom `uiSchema` for the grid to it.

heath-freenome commented on package.json:
```diff
@@ -77,5 +78,9 @@
     "packages/snapshot-tests"
-  ]
+  ],
+  "packageManager": "yarn@1.22.22+sha512.a6b2f7906b721bba3d67d4aff083df04dad64c399707841b7acf00f6b133b7ac24255f2652fa22ae3534329dc6180534e98d17432037ff6fd140556e2bb3137e",
+  "dependencies": {
+    "tailwindcss": "^4.0.1"
+  }
```
Please remove this and do what it takes to have it work with standard `npm install` commands.

heath-freenome commented on packages/daisyui/package.json:
```diff
@@ -0,0 +1,42 @@
+  "version": "6.0.0-alpha.1",
+  "description": "Daisy UI components for react-jsonschema-form",
+  "main": "src/index.ts",
+  "types": "src/index.ts",
+  "scripts": {
```
Can you add these to improve ESM support, matching the changes made in `5.24.7`
```suggestion
  "types": "src/index.ts",
  "exports": {
    ".": {
      "require": "./dist/index.js",
      "import": "./lib/index.js",
      "types": "./lib/index.d.ts"
    },
    "./lib": {
      "require": "./dist/index.js",
      "import": "./lib/index.js",
      "types": "./lib/index.d.ts"
    },
    "./lib/*.js": {
      "require": "./dist/*.js",
      "import": "./lib/*.js",
      "types": "./lib/*.d.ts"
    },
    "./dist": {
      "require": "./dist/index.js",
      "import": "./lib/index.js",
      "types": "./lib/index.d.ts"
    },
    "./dist/*.js": {
      "require": "./dist/*.js",
      "import": "./lib/*.js",
      "types": "./lib/*.d.ts"
    }
  },
  "scripts": {
```

heath-freenome commented on packages/daisyui/package.json:
```diff
@@ -0,0 +1,42 @@
+  "main": "src/index.ts",
+  "types": "src/index.ts",
+  "scripts": {
+    "build": "tsc -b",
```
To improve ESM support you will need to change this to:
```suggestion
    "build:ts": "tsc -b tsconfig.build.json && tsc-alias -p tsconfig.build.json",
```

And add a `tsconfig.build.json` that looks like:
```
{
  "extends": "../../tsconfig.build.json",
  "compilerOptions": {
    "outDir": "./lib"
  },
  "files": [],
  "references": [
    {
      "path": "./src"
    }
  ],
  "tsc-alias": {
    "resolveFullPaths": true,
    "verbose": true,
  }
}
```

heath-freenome commented on packages/daisyui/src/daisyForm/index.ts:
```diff
@@ -0,0 +1,2 @@
+export { default } from './DaisyForm';
+export * from './DaisyForm';
```
```suggestion
export * from './DaisyForm';

```

heath-freenome commented on packages/daisyui/src/templates/ArrayFieldDescriptionTemplate/ArrayFieldDescriptionTemplate.tsx:
To fix the errors on this page, it seems like you need a `tsconfig.json` in your `packages/daisyui` that looks like:
```
{
  "extends": "../../tsconfig.base.json",
  "files": [],
  "references": [
    {
      "path": "./src"
    },
    {
      "path": "./test"
    }
  ]
}
```

heath-freenome commented on packages/daisyui/src/templates/ObjectFieldTemplate/ObjectFieldTemplate.tsx:
```diff
@@ -0,0 +1,87 @@
+        {canExpand<T, S, F>(schema, uiSchema, formData) && (
+          <div className='flex justify-end'>
+            <AddButton
+              className='btn btn-primary btn-sm'
```
Can you add `id={buttonId<T>(id, 'add')}` to match changes made in other themes in v6

heath-freenome commented on packages/daisyui/src/templates/WrapIfAdditionalTemplate/WrapIfAdditionalTemplate.tsx:
```diff
@@ -0,0 +1,41 @@
+          disabled={disabled || readonly}
+        />
+        {schema.additionalProperties && (
+          <button className='btn btn-danger ml-2' onClick={onDropPropertyClick(label)} disabled={disabled || readonly}>
```
Can you add `id={buttonId<T>(id, 'remove')}` to match changes made in other themes in v6

heath-freenome commented on packages/daisyui/src/templates/ArrayFieldTemplate/ArrayFieldTemplate.tsx:
```diff
@@ -0,0 +1,72 @@
+      {canAdd && (
+        <div className='flex flex-row justify-end'>
+          <AddButton
+            className='btn btn-primary'
```
Can you add `id={buttonId<T>(id, 'add')}` to match changes made in other themes in v6

inchoate commented:
@heath-freenome -- haven't forgotten this, I should be circling back to it shortly. Appreciate the patience.

inchoate commented:
@heath-freenome I'm back at it. I'll be rebasing and adapting to the current state of https://github.com/rjsf-team/react-jsonschema-form/tree/rjsf-v6.

Please let me know if I should be doing something else.

heath-freenome commented:
@inchoate I noticed you now have two PRs?

inchoate commented:
@heath-freenome yes -- I left this open for the conversation history. We can close it. Please see the other, which is quite better done.
